File: kAndr/ch_3/count_bits.c
#include <stdio.h>

int count_bits(int x) {
    int num_bits = 0;
    while( x > 0) {
        num_bits += x & 1;
        x >>= 1;
    }
    return num_bits;
}

int main() {
    int x1, x2, x3;
    x1 = count_bits(5);
    x2 = count_bits(4);
    x3 = count_bits(255);

    printf("count_bits of 5 is %d \n", x1);
    printf("count_bits of 4 is %d \n", x2);
    printf("count_bits of 255 is %d \n", x3);
}

File: kAndr/ch_3/count_bits.py
# return how many bits are set to 1 in a positive integer
def count_bits(x: int) -> int:
    num_bits = 0
    while x:
        num_bits += x & 1 # ANDS x against 1 
        x >>= 1 # Binary right shift
    return num_bits

# example of count_bits:
# count_bits(5)
# 101 & 1 = 1
# 10 & 1 = 0
# 1 & 1 = 1
# returns 2

assert count_bits(5) == 2
assert count_bits(4) == 1
assert count_bits(255) == 8